### DISCUSSION TITLE
Announcement regarding PSM OR and inheritance according to #95

### DISCUSSION THREAD

**Discussion by sstenchlak** [mff-uk/dataspecer#179] [Ideas]
This discussion is intended to provide an overview for the plan on how "OR" construct in PSM is to be implemented and how it can handle the inheritance problem specified in #95.

Three new PSM constructs are introduced here. DataPsmInclude, (or include shortly) DataPsmSpecializationOr and DataPsmGeneralizationOr.

### Include

The purpose of DataPsmInclude is to include attributes and associations from one class (or a container in general) to another class as a psmPart next to the class' own attributes and associations. This should correspond to already defined construct psm:Includes used for example in

https://github.com/opendata-mvcr/model-driven-data/blob/2305a4b710e314480e6d2aeb1f252d460dc0bea0/test/psm-rpp-org%C3%A1ny-ve%C5%99ejn%C3%A9-moci.ttl#L75-L78

DataPsmInclude can be placed into classes, next to its attributes and associations. To keep the relations between the interpreted constructs valid, one of the following constraints must be fulfilled:

- Included class must not have an interpretation. (In that case, I suppose it has no interpreted children, and therefore, it is safe to include it.)
- Interpretation of the included class must be an ancestor-or-self of the interpretation of the parent class of the DataPsmInclude. - A set of included attributes and associations must be a subset of possible valid attributes and associations of the parent class.

### OR

There is a need for PSM construct in schema specification that says that "validated data must use exactly one thing from the list of things." DataPsmSpecializationOr chooses a class from a given set of DataPsmClasses and might be used as a root of the tree, or with DataPsmAssociationEnd.

Semantically, this type of OR has a type corresponding to the nearest predecessor of all classes below it. Therefore, all classes below are a specialization of it.

This can be nicely seen with an association end. If the association end has an interpretation, then the class' interpretation must be a descendant-or-self of the corresponding PIM class. By using DataPsmSpecializationOr as the end of the association, we can use its semantic type to enforce only classes that are valid in context of the association.

The example shows the Tourist destination and Sports facility as valid ends of the association and its combination into OR.

![01](https://user-images.githubusercontent.com/23003372/159857972-3684466b-0e6c-46a4-bce2-abcfa4e651a7.svg)

A similar situation may happen at the root of the tree, as a data designer may want to have a schema describing data of different types.

![02](https://user-images.githubusercontent.com/23003372/159858671-42ad9a2f-5cbb-4f1e-a2e1-82ba6eeba9de.svg)

These two usages should be sufficient for every kind of schema possible. However, using only these two may result in unnecessarily complex trees for some specific situations, mainly regarding the cartesian product of multiple ORs in a single class.

Suppose the goal is to model the class with the following interface.
_The class contains a1 only, or b1 and c1 simultaneously; then a2 only, or b2 and c2 simultaneously._
With only constructs introduced so far, the tree would have to contain four classes, each with a different combination of attributes.

![03](https://user-images.githubusercontent.com/23003372/159861180-f029e535-8c6c-4210-b6c8-3eec943b45a6.svg)

It may seem that the solution might be to use two includes, each for one group, as it is seen in the following example. Please note, that include extracts the content of the class below, therefore the combination include-or extracts the context of one class only and behaving as or. For this specific example, the second solution may seem too complex, but for more those independent groups of ors, the former solution would require an exponential number of  classes.

![04](https://user-images.githubusercontent.com/23003372/159861237-8854f311-60a7-4668-9534-ee52f2d93620.svg)

Unfortunately, this solution is invalid, because the include requires ancestor-or-self of its parent, whether the DataPsmSpecializationOr allows descendant-or-self of its internal type. Therefore, for this type of situation, the DataPsmGeneralizationOr is used, permitting only ancestor-or-self classes.

### Are those ORs really different?
The difference is only from a semantic point of view. They behave identically in generators and need to be distinguished only for validation purposes. It is possible to derive which OR is used from the context because they can't be interchanged freely. (Except for the root, where both kinds of OR are theoretically valid) Nonetheless, I would still prefer to distinguish them in the model because of their different semantic meaning.

### Is there any other solution?
We have considered plenty of alternatives. Most of them were just slightly different variants of the proposed solution.

Solving this problem for root seems to be a good starting point. To allow having data of different root classes, there must be root-OR-classes hierarchy.

### Inheritance #95
Suppose we already have the following PSM tree as shown in the image below on the left.

When we decide to "allow, that the Public place may also be a Tourist destination," the following will happen (These steps are shown in the box below):
1. DataPsmSpecializationOr in inserted between the Operates association end and Public place class.
2. PSM class Tourist destination is created.
3. Tourist destination includes the Public place.

![05](https://user-images.githubusercontent.com/23003372/159863879-bad74ee4-c873-461b-8a43-c933ea94eed3.svg)

Additional attributes and association can be added to the Tourist destination class.

These changes will happen on the PSM level. User interface tries to interpret these constructs in a way specified in #95, but in general use of OR and include, these constructs will be shown as are in the PSM model.

**Comment by martinnec** [comment]
Just thinking. Maybe there is much simpler solution. See the end of [this](https://main--kodi-schema-generator.netlify.app/?data-specification=https%3A%2F%2Fofn.gov.cz%2Fdata-specification%2F2203eabe-9807-47e9-8a95-4b7f1ae93267&data-psm-schema=https%3A%2F%2Fofn.gov.cz%2Fschema%2F1643565080290-6284-8e87-84c9&backlink=https%3A%2F%2Fkodi-specification-manager.netlify.app%2Fspecification%3FdataSpecificationIri%3Dhttps%253A%252F%252Fofn.gov.cz%252Fdata-specification%252F2203eabe-9807-47e9-8a95-4b7f1ae93267) data structure. It is weaker but maybe better understandable for a regular data schema designer.
